### ISSUE TITLE
Unsure about the intended behaviour of E0030

### ISSUE THREAD

**Issue by braw-lee** [rust-lang/rust#120014]
[godbolt link](https://godbolt.org/z/sThP53WbG)

```rust
pub fn main() {
    let _a = 1..=0;

    match 0 {
        1..=0 => {},
        _ => {}
    }

    let (1..=0, _b) = (1, 0);
}
```

rustc generates error for empty range pattern in match-like statements only.
Line 2 uses an empty range pattern but is not reported as an error.
Is this the intended behaviour or a bug?

**Comment by ChrisDenton** [comment]
Well it is a slightly different issue. E0030 is about pattern matching whereas line 2 is just constructing a `RangeInclusive`. So

```
let _a = 1..=0;
```
is equivalent to:
```
let _a = RangeInclusive::new(1, 0);
```

This could probably be linted against too although maybe someone has a reason to want it. I can't think why though.

**Comment by ChrisDenton** [comment]
Oh clippy does have a lint for this: https://rust-lang.github.io/rust-clippy/master/index.html#/reversed_empty_ranges

You can try it on the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b9ce1b97ce7633e787b6ed2e3f115735) (open the Tools menu and select Clippy)

**Comment by Nilstrieb** [comment]
In match, both bounds must be known statically. In the expression, they may only be known dynamically.

**Comment by braw-lee** [comment]
thanks for clarification
